### DISCUSSION TITLE
What's the best way to deploy a database?

### DISCUSSION THREAD

**Discussion by Yoric** [diesel-rs/diesel#2721] [Q&A] [answered]
I'm developing an application that uses diesel and, as many young applications, it has a schema that keeps changing:

1. My code receives a `DATABASE_URL` provided by the admins;
2. When my code launches, it connects to the database and runs:
```rust
for source in [include_str!("migrations/migration_1/up.sql"), ..., include_str!("migrations/migration_n/up.sql")] {
  for statement in source.split(';') { // Can't execute multiple statements
    if source.is_empty() {             // Can't execute empty lines
      continue;
    }
    let sql = sql_query(source);
    sql.execute(&connection)?;
  }
}
```

While this kinda works, this has a number of drawbacks:
- the `;` split is very fragile (so, no sql functions for instance, or anything that involves control flow - which makes some migrations really complicated);
- more than once, I've made a mistake in a sql file that somehow went undetected in dev/CI and blew up in the `sql_query`/`sql.execute` lines above in production, which made my life complicated.

Is there any better workflow?

Note that I'm not a db admin, so the only way I can affect the database is through my app.

**Comment by weiznich** [comment, marked as answer]
You likely want to use [`embed_migration!`](https://docs.diesel.rs/1.4.x/diesel_migrations/macro.embed_migrations.html) instead

**Reply by Yoric** [reply]
Thanks!
